src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncDatasetPublishServiceImpl.java

```diff
@@ -3,75 +3,123 @@
 import org.springframework.amqp.core.AmqpTemplate;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Service;
 
 @Service
 class AsyncDatasetPublishServiceImpl implements AsyncDatasetPublishService {
 
-    private static final Logger LOGGER = LoggerFactory
-            .getLogger(AsyncDatasetPublishServiceImpl.class);
-
-
-    private final AmqpTemplate amqpTemplate;
-    private final String createdQueue;
-    private final String transformationToEdmExternalQueue;
-    private final Executor asyncDatasetPublishServiceTaskExecutor;
-
-    public AsyncDatasetPublishServiceImpl(AmqpTemplate amqpTemplate,
-                                          String createdQueue, String transformationToEdmExternalQueue,
-                                          Executor asyncDatasetPublishServiceTaskExecutor) {
-        this.amqpTemplate = amqpTemplate;
-        this.createdQueue = createdQueue;
-        this.asyncDatasetPublishServiceTaskExecutor = asyncDatasetPublishServiceTaskExecutor;
-        this.transformationToEdmExternalQueue = transformationToEdmExternalQueue;
-    }
-
-    @Override
-    public CompletableFuture<Void> publishWithoutXslt(Dataset dataset) {
-        requireNonNull(dataset, "Dataset must not be null");
-        checkArgument(!dataset.getRecords().isEmpty(), "Dataset records must no be empty");
-
-        return CompletableFuture.runAsync(() -> dataset.getRecords()
-                .forEach(this::publishToCreatedQueue), asyncDatasetPublishServiceTaskExecutor);
-    }
-
-    @Override
-    public CompletableFuture<Void> publishWithXslt(Dataset dataset) {
-        requireNonNull(dataset, "Dataset must not be null");
-        checkArgument(!dataset.getRecords().isEmpty(), "Dataset records must no be empty");
-
-        return CompletableFuture.runAsync(() -> dataset.getRecords()
-                .forEach(this::publishToTransformationToEdmExternalQueue), asyncDatasetPublishServiceTaskExecutor);
+  private static final Logger LOGGER = LoggerFactory.getLogger(
+      AsyncDatasetPublishServiceImpl.class);
+
+  @Value("${sandbox.dataset.max-size}")
+  private int maxRecords;
+  private final AmqpTemplate amqpTemplate;
+  private final String createdQueue;
+  private final String oaiHarvestedQueue;
+  private final String transformationToEdmExternalQueue;
+  private final Executor asyncServiceTaskExecutor;
+
+
+  public AsyncDatasetPublishServiceImpl(AmqpTemplate amqpTemplate,
+      String oaiHarvestedQueue, String createdQueue, String transformationToEdmExternalQueue,
+      Executor asyncServiceTaskExecutor) {
+    this.amqpTemplate = amqpTemplate;
+    this.createdQueue = createdQueue;
+    this.oaiHarvestedQueue = oaiHarvestedQueue;
+    this.transformationToEdmExternalQueue = transformationToEdmExternalQueue;
+    this.asyncServiceTaskExecutor = asyncServiceTaskExecutor;
+  }
+
+
+  @Override
+  public CompletableFuture<Void> harvestOaiPmh(String datasetName, String datasetId,
```

Comment:
I'll try to elaborate my thoughts here, but I think it applies from here onwards.
In this method we are creating a "fake" record, we are then sending to a queue harvest information and not records which the queues are intended for, effectively "hacking" the previous concept.
The HarvestOaiPmhExecutor is now not the same type of executor like everything else.

Those observations bring me to the question.
Do we need an oai queue in the first place? The harvested of records could be asynchronously performed here and be sent one by one to the created/xslt queue?

Reply:
This is a good observation. The main reason why we created a oai queue was so we wouldn't wait for the harvesting to be done and for it to start on its own in the "background". This way it will ensure that the user will get an answer fast.

My question now is, if we were to follow your suggestion ("The harvested of records could be asynchronously performed here and be sent one by one to the created/xslt queue?") would we be able to provide an answer to the user fast? Or would we have to wait to finish sending all of the record messages to the queues?

Reply:
Yes it should should be asynchronous.
Essentially what is done in HarvestOaiPmhExecutor could be done here calling a method from inside the asynchronous executor `CompletableFuture.runAsync()`. Similarly to the rest.
So the reply will be as fast as like that too.

Reply:
I've tried to test this suggestion after implementing this suggestion (or at least something similar to it) and we do *not* get an immediate answer to the user because we end up waiting for the loop (the one you see in HarvestOaiPmhExecutor) to end, even though we are sending sending each harvested record to a queue. I guess this now depends on the preference of how fast the response should be